specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,134 @@
+# bought in a block. Calling code must handle the gas burn and validity checks on
+# the ability of the account to afford this gas.
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+GUARANTEED_GAS_LIMIT = 2,500,000
+GAS_TARGET = GUARANTEED_GAS_LIMIT / ELASTICITY_MULTIPLIER
+    
+# All values are uint64s
+prev_basefee, prev_num, prev_bought_gas = load_and_unpack_storage()
+now_num = block.number
+
+# Clamp the full basefee to a specific range. The minimum value in the range should be around 100-1000
+# to enable faster responses in the basefee. This replaces the `max` mechanism in the ethereum 1559
+# implementation (it also serves to enable the basefee to increase if it is very small).
+def clamp(v: i256, min: u64, max: u64) -> u64:
+    if v < i256(min):
+        return min
+    elif v > i256(max):
+        return max
+    else:
+        return u64(v)
+
+# If this is a new block, update the basefee and reset the total gas
+# If not, just update the total gas
+if prev_num == now_num:
+    now_basefee = prev_basefee
+    now_bought_gas = prev_bought_gas + requested_gas
+elif prev_num != now_num :
+    # New formula
+    # Width extension and conversion to signed integer math
+    gas_used_delta = int128(prev_bought_gas) - int128(GAS_TARGET)
+    # Use truncating (round to 0) division - solidity's default.
+    # Sign extend gas_used_delta & prev_basefee to 256 bits to avoid overflows here.
+    base_fee_per_gas_delta = prev_basefee * gas_used_delta / GAS_TARGET / BASE_FEE_MAX_CHANGE_DENOMINATOR
+    now_basefee_wide = prev_basefee + base_fee_per_gas_delta
+
+    now_basefee = clamp(now_basefee_wide, min=1000, max=UINT_64_MAX_VALUE)
+    now_bought_gas =  requested_gas
+
+# If we skipped multiple blocks between the previous block and now update the basefee again.
+# This is not exactly the same as iterating the above function, but quite close for reasonable
+# gas target values. It is also constant time wrt the number of missed blocks which is important
+# for keeping gas usage stable.
+if prev_num + 1 < now_num:
+    n = now_num - prev_num - 1
+    # Apply 7/8 reduction to prev_basefee for the n empty blocks in a row.
+    # TODO: Make sure this doesn't overflow
+    now_basefee_wide = prev_basefee * 7**n / 8**n
+    now_basefee = clamp(now_basefee_wide, min=1000, max=UINT_64_MAX_VALUE)
+
+require(now_bought_gas < GUARANTEED_GAS_LIMIT)
+
+
+pack_and_store(now_basefee, now_num, now_bought_gas)
+```
+
+### Exponent Based Fee Reduction
+
+When there are stretches where no deposits are executed on L1, the basefee should be decaying, but
+is not. If there is the case that the basefee spiked, this mechanism is needed to enable a more
+accurate decay. It uses exponentiation to run in constant (relative to the number of missed blocks)
+gas.
+
+With the current elasticty and change denominator values, if `n` is greater than 42, the
+exponention will overflow. TODO: Use exponent/logarithm library for this.
```

Comment:
If `n > 42` then there have been no deposits for a long period of time `42 * 12 = 504 seconds = 8.4 minutes`. At that point we may as well reset the base fee for deposits to a default minimum value. No need for special libraries that way.

Reply:
`(7/8)^42 ~1e-3` which is a large reduction in basefee, but it doesn't feel large enough to reset it to the minimum. It is something that I thought about.

Reply:
Definitely an option. This isn't fully decided yet